README
NAME
  Ado::Manual - Developers' manual

DESCRIPTION
  Ado is a light on dependencies framework and application for web-projects,
  based on Mojolicious, written in the Perl programming language
  <http://www.perl.org/>. Ado is a typical well structured, MVC
  <http://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller>
  Mojolicious application. It aims to allow teams to build on it a CMS, ERP,
  CRM or all of them integrated together.

  Please note that the project is still a work-in-progress. Parts of it
  work, others are simply missing.

GETTING STARTED
  *   Start a personal blog with Ado::Manual::FiveMinutes if you are
      impatient.

  *   Read the Ado::Manual::Intro for a general overview.

  *   To see different installation methods, go to
      Ado::Manual::Installation.

  *   To learn about plugins and how to write your own plugin, see
      Ado::Manual::Plugins.

SUPPORT AND DOCUMENTATION
  After installing, you can find documentation with the perldoc command.

          perldoc Ado
          perldoc Ado::Manual #this page

  For better experience run the "ado" application and read the documentation
  from your browser.

          /path/to/yourAdo/bin/ado daemon

  If you installed "ado" in your Perl distro, you can simply type:

          ado daemon

  Go to http:/localhost:3000/perldoc

  You can report bugs and suggest features at
  <http://github.com/kberov/Ado/issues>. Bugs will be considered and fixed
  as time permits. Feel invited to make pull requests for your
  contributions.

  If you are simply looking for help with using Ado, please ask your
  questions at <https://groups.google.com/d/forum/ado-dev>.

CONTRIBUTING
  Anybody can contribute by reporting issues via github or fixing typos in
  the documentation. To be able to contribute with code, some rules need to
  be kept. This is mandatory for any community project. Generally the rules
  outlined in Mojolicious::Guides::Contributing apply for Ado too. For
  specific to Ado rules see Ado::Manual::Contributing.

  We expect that you know how Internet works, how to write Perl modules and
  are familiar with Mojolicious.

  To ease discussions on Ado further development, a forum was created.
  <https://groups.google.com/d/forum/ado-dev>.

REST API
  Ado strives for strict separation of concerns. The best way to achieve
  this is to fully separate the client code from the server code. Ado is
  ideally suited for the purpose thanks to Mojolicious. Every
  resource(route) is accessible via a browser as "/path/to/resourse" an
  returns HTML or using "/path/to/resourse.json" and returns JSON. We follow
  closely and elaborate on the recommendations in "RESTful Service Best
  Practices" at <http://www.RestApiTutorial.com>. See Ado::Manual::RESTAPI.

PLUGINS
  Business-specific applications for an Ado-based system are usually
  implemented as plugins. One way to contribute to Ado is by writing
  plugins.

  Ado plugins work the same way as Mojolicious::Plugins and share the same
  common base trough Ado::Plugin which ISA Mojolicious::Plugins. Ado plugins
  have one small additional feature. They can load their own configuration
  from "$ENV{MOJO_HOME}/etc/plugins/plugin_name.conf".

  See Ado::Manual::Plugins and Ado::Plugin for more information.

CONTINUOUS INTEGRATION
  We would like to know that our software is always in good health. We count
  on friendly developers and organizations to install and test it
  continuously.

  CPAN Testers Reports for Ado
  <http://www.cpantesters.org/distro/A/Ado.html>

  Travis-CI <https://travis-ci.org/kberov/Ado>

SEE ALSO
  Ado, Mojolicious::Guides, Mojolicious::Guides::Contributing,
  <http://www.thefreedictionary.com/ado>.

AUTHORS
  Authors, ordered by contributions
  (<https://github.com/kberov/Ado/graphs/contributors>).

  Красимир Беров (Krasimir Berov)(berov@cpan.org)

  Вълчо Неделчев (Valcho Nedelchev)(kumcho@vulcho.com)

  Joachim Astel

  Renee Baecker (module@renee-baecker.de)

COPYRIGHT AND LICENSE
  Copyright 2013-2015 Красимир Беров (Krasimir Berov).

  This program is free software, you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License v3 (LGPL-3.0).
  You may copy, distribute and modify the software provided that
  modifications are open source. However, software that includes the license
  may release under a different license.

  See http://opensource.org/licenses/lgpl-3.0.html for more information.



Changes
Revision history for Ado

0.935 2017-09-02
  - Upgraded to Mojolicious 7.43.
  - Now trying to use ado.conf and ado.sqlite from $app->ado_home
    if the files are not found in $app->home.

0.934 2017-06-17
  - Upgraded to Mojolicious 7.33.
  - Upgraded to Perl::Critic 1.28 and Test::Perl::Critic 1.03.

0.933 2016-12-05
  - Fixed occasionally failing test in t/basic.t.
  - Required some higher versions for recommended modules.

0.932 2016-12-04
  - Upgraded to Mojolicious 7.11, Mojolicious::Plugin::OAuth2 1.53
    and probably others.
  - Fixed failing test in t/basic.t.
  - Cleared trailing spaces - RTKH (Artur Khabibullin).
  - Made some sub arguments assignment slightly more
    efficient - KABANOID (Michael Katasonov).

0.931 2016-07-06
  - Upgraded to Mojolicious 7.01.
  - Fixed some failing tests with the new Mojo.
  - Unnoticeable pod reformatting.

0.930 2016-04-15
  - Upgraded to Mojolicious 6.58.
  - Expanded templates from Ado::Plugin::Auth.
  - Fixed bugs related to improper dynamic routes names usage.
  - $CODENAME changed to "ИЖЕ" U+2C0A GLAGOLITIC CAPITAL LETTER INITIAL IZHE (Ⰺ).

0.921 2015-06-03
  - Fixed query from i18n table to retrieve translated messages.
  - Rearranged Plugin::MarkdounRender tests to avoid potential failures.

0.92 2015-06-01
  - Ado::Control::list_for_json() accepts a new argument $meta for arbitrary
  content, specific to the served resource.
  - Removed a lot of debug messages.
  - Added database storage for lexicons to Ado::I18n.
    Now plugins can store their translated messages in i18n table.
    See Ado::Plugin::Vest as example.
  - Refactored and improved /articles section.
  - Decided not to remove documents from public/doc. They will be used as
    end-user documentation and written when time permits.
  - $CODENAME changed to "иже" U+2C09 GLAGOLITIC CAPITAL LETTER IZHE (Ⰹ)

0.91 2015-05-10
  - Upgraded to Mojolicious::Plugin::SemanticUI 0.17 to use only some of
    Semantic UI components. This way the first rendering of the default
    page is faster on the browser side.
  - Implemented header_css and header_javascript helpers to add minimalistic asset
    management. Used to refer specific Semantic UI components from templates
    and render links to them in the <head> section of the default layout.
  - The /help section is deprecated and will be removed in the next release.
  - Other cleanups...

0.901 2015-05-01
  - Fixed bug in Ado::Plugin::Auth. Now you can disable local authentication
    and use only OAuth2 authentication methods.
  - Tested and works with Mojolicious 6.10.
  - Cleaned up some (now seeming meaningless) TODOs.

0.90 2015-04-26
  - Using $app->moniker instead of hardcodded 'ado' in
    Ado::Command::generate::apache2htaccess.
  - Allow when Ado is installed globally application specific plugin
    configuration to override the global plugin configuration.
  - Improved test coverage of Ado, Ado::Plugin and
    Ado::Command::generate::adoplugin. Total: 81.7% with TEST_AUTHOR set.

0.89 2015-04-19
  - Added hard dependency to Mojolicious::Plugin::OAuth2.
  - Upgraded to Mojolicious::Plugin::OAuth2 1.52.
  - Upgraded to Mojolicious 6.08.
  - Added method by_email to Ado::Model::Users.

0.88 2015-04-18
  - Updated Ado::Manual::Intro Ado::Manual::Plugins and Ado::Manual::Installation.
    Created Ado::Manual::FiveMinutes.
  - Ado::Plugin::MarkdownRenderer also makes its configuration available for
    usage at runtime by controller classes.
  - Implemented Ado::Control::Articles specifically for Ado::Manual::FiveMinutes.

0.87 2015-03-30
  - Moved table domains to Ado::Plugin::Site.
  - Improved Error reporting in do_sql_file in Ado::Plugin::AdoHelpers.
  - Fixed bug in Ado::Sessions. It was failing to set default session type.

0.86 2015-03-30
  - Started work in Ado::Build to handle upgrades beside installations
    so the sqlite database is not replaced with the newly installed version
    and thus destroing content of unconscious users - Alpha.
  - Writing Ado::Manual::Intro.
  - Upgraded to Mojolicious 6.05!

0.85 2015-03-22
  - Added hook 'after_login', emited by Ado::Plugin::Auth.
  - Added possibilty to add items to the adobar menu
    via session->{adobar_links}.
  - $CODENAME changed to
    "земля" - U+2C08 GLAGOLITIC CAPITAL LETTER ZEMLJA (Ⰸ)
  - Showing current controller and action only when in development mode
    (templates/layouts/default.html.ep).
  - Upgraded to Mojolicious::Plugin::SemanticUI 0.15!
  - Upgraded to DBIx::Simple::Class 1.009!
  - Upgraded to Mojolicious::Plugin::DSC 1.005!
  - Upgraded to Mojolicious 6.03!
  - Aded 'query' - hash reference - containing the parameters in Ado::Control::list_for_json().

0.84 2015-03-13
  - Requiring Mojolicious 6.0 explicitly.
  - Using Perl::Critic 1.124. Simplified t/.perlcriticrc.
  - Fixed bug in comands apache2htaccess and apache2vhost.
    They were failing to find their templates.
    Feedback by Francisco Ortiz-Muñoz. Thanks Francisco!

0.83 2015-02-27
  - Continued with cleanups. New changes are applyed as need
    arises while working on Ado::Plugin::Admin.
  - Upgraded to Mojolicious 6.0!


0.82 2015-02-24
- Improved Ado::_initialise.
- Adding paths to bundled files only in case Ado is used as a
  library. Otherwise Mojo does it anyway.


0.81 2015-02-24
  - Implemented Ado::ado_home - pointing always to where
    Ado is installed.
  - Moved default initialisation from ado.conf to Ado::_initialise.
  - Improved Ado::Manual.

0.80 2015-02-22
  - Implemented do_sql_file application helper in Ado::Plugin::AdoHelpers.
  - Improved documentation.
    Added the newly created google group Ado-dev to Ado::Manual page.
  - Upgraded to Mojolicious 5.81.
  - Upgraded to Mojolicious::Plugin::DSC 1.002.

0.79 2015-02-15
  - $ENV{MOJO_HOME} is not set explicitly any more.
    Detecting properly app->home and falling back to Mojo::home.
  - Cleanups.
    - Moved Ado::Control::Ado to Ado-Plugin-Admin where it logically belongs.

Build.PL
use 5.014000;
use strict;
use warnings FATAL => 'all';
use FindBin;
use lib("$FindBin::Bin/lib");
use Ado::Build;

#unset temporarily $ENV{TEST_AUTHOR}
#to generate fewer dependencies dynamically
#see http://localhost:3000/perldoc/CPAN/Meta/Spec#dynamic_config
my $author = {
    build_requires => {

        #Modules for QA - while developping
        'Perl::Tidy'          => '20140711',
        'Test::Output'        => '1.03',
        'Perl::Critic'        => '1.124',
        'Test::Perl::Critic'  => '1.02',
        'Test::CheckManifest' => '1.28',
        'Test::Pod::Coverage' => '1.10',
        'Test::Pod'           => '1.48',
        'Lingua::Ispell'      => '0',
        'Test::Pod::Spelling' => '0.4',
        'Devel::Cover'        => '1.15',
        'Test::Exception'     => '0.32',
        'Test::Differences'   => '0.62',
        'Pod::Markdown'       => '2.002',
    },
    configure_requires => {},
    requires           => {

        #To develop and test advanced functionality
        #This should go as "recommends" for regular users
        #when we have functionality depending on these
        'IO::Socket::SSL' => '2.039',
        'EV'              => '4.22',
        'IO::Socket::IP'  => '0.37',

        #'DBD::mysql'      => '4.025',
    },
};
my $builder = Ado::Build->new(
    module_name => 'Ado',
    license     => 'lgpl_3_0',
    $ENV{TEST_AUTHOR} ? (create_readme => 1) : (),
    dist_author => q{Красимир Беров <berov@cpan.org>},

    #release_status     => 'unstable',
    configure_requires => {
        'Module::Build' => '0.42',
        $ENV{TEST_AUTHOR} ? (%{$$author{configure_requires}}) : (),
    },
    build_requires => {
        'Module::Build' => '0.42',
        'Test::More'    => '0',
        $ENV{TEST_AUTHOR} ? (%{$$author{build_requires}}) : (),
    },
    requires => {
        'perl'                            => '5.14.0',
        'Mojolicious'                     => '7.43',
        'Mojolicious::Plugin::DSC'        => '1.005',
        'Mojolicious::Plugin::SemanticUI' => '0.17',
        'Mojolicious::Plugin::OAuth2'     => '1.52',
        'DBIx::Simple::Class'             => '1.009',
        'Email::Address'                  => '1.905',
        'Locale::Maketext'                => '1.25',
        $ENV{TEST_AUTHOR} ? (%{$$author{requires}}) : (),

        #Default markdown renderer
        'Text::MultiMarkdown' => '1.000035',

    },
    recommends => {

        'DBD::SQLite'   => '1.52',
        'SQL::Abstract' => '1.78',

        # Either
        'Plack'                       => '1.0030',
        'FCGI'                        => '0.74',
        'FCGI::ProcManager'           => '0.24',
        'Apache::LogFormat::Compiler' => '0',

        # or
        'Mojo::Server::FastCGI' => '0.41',
    },
    recursive_test_files => 1,
    meta_merge           => {
        resources => {
            repository => "https://github.com/kberov/Ado",
            bugtracker => "https://github.com/kberov/Ado/issues",
        },
        keywords => [qw/Ado Mojolicious realtime web ERP REST CMS enterprise/],
        no_index => {namespace => ['Ado::Model'],}
    },
    script_files   => 'bin',
    add_to_cleanup => [
        'Ado-*',                     '*.bak',
        'public/doc/**/*.html',      'public/articles/*.html',
        'public/articles/**/*.html', 't/ado/etc/ado.sqlite'
    ],
);

$builder->create_build_script();

